pkg/distributor/aggregator/aggregator.go

```diff
@@ -0,0 +1,278 @@
+package aggregator
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// Aggregator aggregates values within
+// a time window over a period of time.
+type Aggregator[T any] struct {
+	window int64
+	period int64
+	now    func() int64
+
+	m          sync.RWMutex
+	tracker    *tracker
+	aggregates map[aggregationKey]*aggregatedResult[T]
+
+	close chan struct{}
+	done  chan struct{}
+	stats stats
+}
+
+type stats struct {
+	activeAggregates atomic.Int64
+	activeSeries     atomic.Uint64
+	aggregated       atomic.Uint64
+	errors           atomic.Uint64
+}
+
+func NewAggregator[T any](window, period time.Duration) *Aggregator[T] {
+	if window < period {
+		window = period
+	}
+	return &Aggregator[T]{
+		window:  window.Nanoseconds(),
+		period:  period.Nanoseconds(),
+		now:     timeNow,
+		tracker: newTracker(8, 64),
+		// NOTE(kolesnikovae): probably should be sharded as well.
+		aggregates: make(map[aggregationKey]*aggregatedResult[T], 256),
+		close:      make(chan struct{}),
+		done:       make(chan struct{}),
+	}
+}
+
+func timeNow() int64 { return time.Now().UnixNano() }
+
+func (a *Aggregator[T]) Start() {
+	t := time.NewTicker(time.Duration(a.period))
+	defer func() {
+		t.Stop()
+		close(a.done)
+	}()
+	for {
+		select {
+		case <-a.close:
+			return
+		case <-t.C:
+			a.prune(a.now())
+		}
+	}
+}
+
+// Stop the aggregator. It does not wait for ongoing aggregations
+// to complete as no aggregation requests expected during shutdown.
+func (a *Aggregator[T]) Stop() {
+	close(a.close)
+	<-a.done
+}
+
+type AggregateFn[T any] func(T) (T, error)
```

Comment:
This could be also used during compaction for doing sampling, really like the direction of that code.